corehq/motech/value_source.py

```diff
@@ -143,7 +145,71 @@ def check_direction(self, direction):
         return not self.direction or direction == self.direction
 
 
-class CaseProperty(ValueSource):
+class MapMixin(DocumentSchema):
+    """
+    Maps CommCare values to OpenMRS values or concept UUIDs
+
+    Example "person_attribute" value::
+
+        {
+          "00000000-771d-0000-0000-000000000000": {
+            "doc_type": "CasePropertyMap",
+            "case_property": "pill"
+            "value_map": {
+              "red": "00ff0000-771d-0000-0000-000000000000",
+              "blue": "000000ff-771d-0000-0000-000000000000",
+            }
+          }
+        }
+
+    """
+    value_map = DictProperty(required=False, default=None, exclude_if_none=True)
+
+    def serialize(self, value):
+        """
+        Uses self.value_map to serialize instead of converting data
+        types as ``ValueSource`` does.
+        """
+        if self.value_map is None:
+            return super().serialize(value)
+        # Use `.get()` because it's OK if some CommCare answers are
+        # not mapped to OpenMRS concepts, e.g. when only the "yes" value
+        # of a yes-no question in CommCare is mapped to a concept in
+        # OpenMRS.
+        return self.value_map.get(value)
+
+    def deserialize(self, external_value):
+        if self.value_map is None:
+            return super().deserialize(external_value)
+        reverse_map = {v: k for k, v in self.value_map.items()}
+        return reverse_map.get(external_value)
+
+
+class JsonPathMixin(DocumentSchema):
+    """
+    Used for importing a value from a JSON document.
+    """
+    jsonpath = StringProperty(required=False, default=None, exclude_if_none=True)
+
+    def _get_external_value(self, external_data):
+        if self.jsonpath is None:
+            return None
+        jsonpath = parse_jsonpath(self.jsonpath)
+        matches = jsonpath.find(external_data)
+        values = [m.value for m in matches]
+        if not values:
+            return None
+        elif len(values) == 1:
+            return values[0]
+        else:
+            return values
+
+    def get_import_value(self, external_data):
+        external_value = self._get_external_value(external_data)
+        return self.deserialize(external_value)
+
+
+class CaseProperty(MapMixin, ValueSource, JsonPathMixin):
```

Comment:
The order of these mixins is intriguing. I'm assuming it's intentional (and important) that `MapMixin` comes before `ValueSource`, and `JsonPathMixin` comes after? This kind of intricate base class ordering can make the MRO rather tricky to reason about.

Reply:
`MapMixin` must come before `ValueSource` so that the MRO finds its `serialize()` and `deserialize()` methods first. The order of `JsonPathMixin` doesn't matter because it doesn't implement the methods or properties of the others.

:thinking: Maybe a comment would help.